[lock/unlock]: support multi-stage finalization in rebalancer contract

makramkd opened the pull request:
## Motivation

PR #606 implements multi-stage finalization in the Optimism L1 bridge adapter, however this is not functional end-to-end because the Rebalancer assumes that finalization happens in a single step.

## Solution

Change the bridge adapter interface to return a boolean from `finalizeWithdrawERC20` that indicates whether the funds are available for use by the rebalancer.

The rest of the changes are to get test coverage for this new interface change.

makramkd committed:
implement OP L1 finalization in bridge adapter

makramkd committed:
prettier, solhint

makramkd committed:
revert comment

makramkd committed:
remove unused var

makramkd committed:
Merge branch 'ccip-develop' into makram/CCIP-1438/opstack-l1-adapter

makramkd committed:
wip

makramkd committed:
test coverage for testonlybridge

makramkd committed:
enforce a prove before a finalize on mock adapter

makramkd committed:
prettier

makramkd committed:
fix bridge data encoding in test only bridge

makramkd committed:
fix tests

makramkd committed:
goimports

makramkd committed:
simplify test

makramkd committed:
pr comments

makramkd committed:
add foundry test for finalizeWithdrawERC20

tests encoding and if logic

makramkd committed:
prettier

makramkd committed:
Merge branch 'makram/CCIP-1438/opstack-l1-adapter' into makram/CCIP-1438/rebalancer-changes

makramkd committed:
bump snapshot

makramkd committed:
Merge branch 'makram/CCIP-1438/opstack-l1-adapter' into makram/CCIP-1438/rebalancer-changes

makramkd committed:
fix unit test

makramkd committed:
add more tests

makramkd committed:
Merge branch 'ccip-develop' into makram/CCIP-1438/rebalancer-changes

makramkd committed:
cleanup comments

dimkouv commented on contracts/src/v0.8/rebalancer/test/Rebalancer.t.sol:
```diff
@@ -258,81 +263,207 @@ contract Rebalancer_rebalanceLiquidity is RebalancerSetup {
     // this should transfer the funds to the rebalancer.
-    bytes memory finalizationData = abi.encode(amount, nonce);
-    s_bridgeAdapter.finalizeWithdrawERC20(address(0), address(s_rebalancer), finalizationData);
+    MockL1BridgeAdapter.ProvePayload memory provePayload = MockL1BridgeAdapter.ProvePayload({nonce: nonce});
```
nice

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l1_to_l2.go:
```diff
@@ -372,6 +373,7 @@ func (l *l1ToL2Bridge) toPendingTransfers(
 					logIndex: int64(transfer.Raw.Index),
 				}].BlockTimestamp,
 				BridgeData: readyData[i], // finalization data since its ready
+				Stage:      2,
```
worths declaring consts

```
const (
   arb.Stage1 // Stage1 is blah blah..
   arb.Stage2 // Stage2 is ...
)
```

makramkd replied:
Agree, will declare and will describe this idea in more detail.

makramkd replied:
There's still a bunch of follow up work here, put it in a ticket here: https://smartcontract-it.atlassian.net/browse/CCIP-1893 so as to not distract from the rest of the PR since its somewhat nontrivial (functions need to be updated, tests, etc.)

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/testonlybridge/l1_to_l1.go:
```diff
@@ -47,9 +59,11 @@ func New(
 ) (*testBridge, error) {
 	err := sourceLogPoller.RegisterFilter(logpoller.Filter{
-		Name: logpoller.FilterName("L1-LiquidityTransferred", sourceSelector),
+		Name: logpoller.FilterName("Local-LiquidityTransferred-FinalizationCompleted",
```
Maybe we can improve the way we manage those filters?
Maybe keep them in a vars group and/or keep them as struct fields.

makramkd replied:
I think eventually we will have to call the `Close` functions on these bridges, right now we don't do that.

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/testonlybridge/l1_to_l1.go:
```diff
@@ -172,26 +188,61 @@ func (t *testBridge) GetTransfers(ctx context.Context, localToken models.Address
+		lp := parsedToLP[logKey{txHash: send.Raw.TxHash, logIndex: int64(send.Raw.Index)}]
+		sendNonce, err := UnpackBridgeSendReturnData(send.BridgeReturnData)
+		if err != nil {
+			t.lggr.Errorw("unpack send bridge data", "err", err)
```
In which case unpack returns err? maybe worths returning the error and stopping?

makramkd replied:
Yeah, I think we can do that.

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/testonlybridge/l1_to_l1.go:
```diff
@@ -227,18 +279,95 @@ func (t *testBridge) toPendingTransfers(
+			"readyToFinalize", readyToFinalize)
+	}
+	if len(readyToProve) > 0 {
+		lggr.Infow("found unproven sends, ready to prove",
```
nice, we need such logs

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/testonlybridge/l1_to_l1.go:
```diff
@@ -227,18 +279,95 @@ func (t *testBridge) toPendingTransfers(
+}
+
+func groupByStage(
+	unfinalized []*rebalancer.RebalancerLiquidityTransferred,
```
why slice of pointers? imo better to not use slice of pointers, might prevent weird bugs

makramkd replied:
The `Parse*` functions generated by `abigen` all return pointers.

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/testonlybridge/l1_to_l1.go:
```diff
@@ -227,18 +279,95 @@ func (t *testBridge) toPendingTransfers(
+) {
+	for _, candidate := range unfinalized {
+		var proven bool
+		for _, stepCompleted := range stepsCompleted {
```
Maybe cleaner to place the nested loop in a function.

dimkouv commented on core/services/ocr2/plugins/rebalancer/inflight/inflight.go:
```diff
@@ -60,8 +60,8 @@ func (i *inflight) Expire(pending []models.PendingTransfer) int {
-		_, ok := i.transfers[k]
-		if ok {
+		t, ok := i.transfers[k]
+		if ok && p.Stage > t.Stage {
```
maybe place a comment here to explain what `p.state > t.state` means

dimkouv reviewed (commented):
lgtm, minor comments

makramkd committed:
code review changes

makramkd committed:
Merge branch 'ccip-develop' into makram/CCIP-1438/rebalancer-changes

makramkd committed:
fix test

makramkd committed:
fix test

makramkd committed:
[lock/unlock]: rewrap eth if flag set (#625)

## Motivation

Optimism ETH transfers need to be rewrapped before we can inject them
into the token pool.

## Solution

Add a parameter to the `ReceiveLiquidityParams` that indicates whether
the Rebalancer contract should re-wrap native after executing a
finalization.

It is expected that the caller knows what they are doing (i.e DON or
owner).

makramkd committed:
Merge branch 'ccip-develop' into makram/CCIP-1438/rebalancer-changes

makramkd committed:
fix test

makramkd committed:
fix struct packing

makramkd committed:
regenerate

makramkd committed:
Merge branch 'ccip-develop' into makram/CCIP-1438/rebalancer-changes

matYang commented on contracts/src/v0.8/rebalancer/test/mocks/MockBridgeAdapter.sol:
```diff
@@ -52,20 +104,49 @@ contract MockL1BridgeAdapter is IBridgeAdapter, ILiquidityContainer {
-    s_nonceUsed[nonce] = true;
-    i_token.safeTransfer(localReceiver, amount);
+  ) external override returns (bool) {
+    Payload memory payload = abi.decode(bridgeSpecificPayload, (Payload));
```
nit: wow! ik it's a mock, but still let's use return early to reduce some nesting

makramkd replied:
Fixed!

matYang commented on contracts/src/v0.8/rebalancer/test/Rebalancer.t.sol:
```diff
@@ -115,9 +144,10 @@ contract Rebalancer__report is RebalancerSetup {
 }
 
 contract Rebalancer_rebalanceLiquidity is RebalancerSetup {
+  uint256 internal constant AMOUNT = 12345679;
```
static AMOUNT implies these are good fuzz candidates, let's add some fuzzing? =)

makramkd replied:
There are potentially opportunities for fuzzing, but I don't think the current tests are useful to make fuzzes. They basically just `deal` this `AMOUNT` to the pool and run through the rebalancing process.

Adding fuzz tests is on the docket though, we have tickets for that.

matYang commented on contracts/src/v0.8/rebalancer/Rebalancer.sol:
```diff
@@ -242,8 +261,29 @@ contract Rebalancer is IRebalancer, OCR3Base {
-    // approve and liquidity container should transferFrom
+    // if we reach this point, the finalization failed.
+    // since we don't have enough information to know why it failed,
+    // we assume that it failed because the withdrawal was already finalized,
```
instead of assuming, let's explain why this will always be safe to call injectLiquidity here

makramkd replied:
This logic will change in upcoming PRs after recent reviews, so stay tuned.

makramkd committed:
simplify mock bridge

makramkd committed:
bump snapshots

makramkd committed:
fix lint

RensR commented on contracts/src/v0.8/rebalancer/Rebalancer.sol:
```diff
@@ -4,15 +4,15 @@ pragma solidity 0.8.19;
 import {IRebalancer} from "./interfaces/IRebalancer.sol";
 import {ILiquidityContainer} from "./interfaces/ILiquidityContainer.sol";
-
+import {IWrappedNative} from "../ccip/interfaces/IWrappedNative.sol";
```
nit newline after interfaces

makramkd replied:
Fixed 👍

RensR commented on contracts/src/v0.8/rebalancer/Rebalancer.sol:
```diff
@@ -258,6 +298,11 @@ contract Rebalancer is IRebalancer, OCR3Base {
   }
 
+  function _wrapNative(uint256 amount) private {
+    IWrappedNative weth = IWrappedNative(address(i_localToken));
```
nit: remove assignment and inline. Might as well remove the function and just inline the single line at that point?

makramkd replied:
Inlined 👍

RensR commented on contracts/src/v0.8/rebalancer/test/mocks/MockBridgeAdapter.sol:
```diff
@@ -52,20 +104,61 @@ contract MockL1BridgeAdapter is IBridgeAdapter, ILiquidityContainer {
+      return _proveWithdrawal(payload);
+    } else if (payload.action == FinalizationAction.FinalizeWithdrawal) {
+      return _finalizeWithdrawal(payload, localReceiver);
+    } else {
```
nit else can be removed

makramkd replied:
Axed 🪓

makramkd committed:
add changesets, fix configs

makramkd committed:
inline deposit call
